deployment/operations/execute.go

```diff
@@ -0,0 +1,82 @@
+package operations
+
+import (
+	"github.com/avast/retry-go/v4"
+)
+
+// ExecuteConfig is the configuration for the ExecuteOp function.
+type ExecuteConfig[Input any] struct {
+	inputHook func(input Input) Input
+	noRetry   bool
+}
+
+type ExecuteOption[Input any] func(*ExecuteConfig[Input])
+
+// WithNewInputHook is an ExecuteOption that performs changes to the input before retrying the operation with
+// the new input. This allows scenario like changing the gas limit option.
+func WithNewInputHook[Input any](hook func(input Input) Input) ExecuteOption[Input] {
+	return func(c *ExecuteConfig[Input]) {
+		c.inputHook = hook
+	}
+}
+
+// WithNoRetry is an ExecuteOption that disables retrying.
+func WithNoRetry[Input any]() ExecuteOption[Input] {
+	return func(c *ExecuteConfig[Input]) {
+		c.noRetry = true
+	}
+}
+
+// ExecuteOp executes an operation with the given input and dependencies.
+// By default, it retries the operation up to 10 times with exponential backoff if it fails.
+// Use WithNewInputHook and WithNoRetry to customize the retry behavior.
+func ExecuteOp[Input, Output, Deps any](
+	env OperationEnv,
+	operation *Operation[Input, Output, Deps],
+	deps Deps,
+	input Input,
+	opts ...ExecuteOption[Input],
```

Comment:
I like the flexibility, but it doesn't protect the developer very well from using multiple, incompatible options. The logic flow does seem to handle it at the moment, because of the no-retry if statement, but it means the control is implicit, and if a user provided both, they'd get no signal they'd made an error in the usage.

I would suggest that these, despite my liking generalization, the current options all relate to retry policy (none, or custom handler). I'd consider not having "Opts" but rather having "retryPolicy" and using this current code, renamed to just be about retries. IF we have more general options, we can evolve it there.

Reply:
The mechanism is fine, it's mostly the use of variadic opts, and calling them general opts, rather than RetryPolicy. The config approach is great.

Reply:
Or actually... are you using variadics to allow someone to say `ExecuteOp(e, o, in, out)` or `ExecuteOp(e, o, in, out, opt)`?  If that's the case, and it's just to sort of simulate function overloading in a way, then I guess that's fine, but I would document the crap about that usage pattern, and raise an error if someone adds in more than one.